iceoryx_posh/source/popo/condition.cpp

```diff
@@ -14,61 +14,59 @@
 
 #include "iceoryx_posh/popo/condition.hpp"
 
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/condition_variable_data.hpp"
+#include "iceoryx_posh/popo/wait_set.hpp"
+#include "iceoryx_utils/error_handling/error_handling.hpp"
+
+
 namespace iox
 {
 namespace popo
 {
 Condition::~Condition() noexcept
 {
-    // WaitSet is still alive as it uses RAII, hence our contract with the user isn't fulfilled anymore (no dangling
-    // condition allowed)
     if (isConditionVariableAttached())
     {
-        errorHandler(Error::kPOPO__WAITSET_CONDITION_LIFETIME_ISSUE, nullptr, ErrorLevel::FATAL);
+        // In this particular case we can skip unsetConditionVariable since in
+        // dtor the object already degraded to a class which is only a condition
+        // and has no more vtable pointer to the method unsetConditionVariable
+        // which was once part of this object
+        // If we assume that
+        m_waitSet.load(std::memory_order_relaxed)->removeCondition(*this);
     }
 }
 
-Condition::Condition(const Condition& rhs) noexcept
-{
-    m_conditionVariableAttached.store(rhs.m_conditionVariableAttached, std::memory_order_relaxed);
-}
-
 bool Condition::isConditionVariableAttached() const noexcept
 {
-    return m_conditionVariableAttached.load(std::memory_order_relaxed);
+    return m_waitSet.load(std::memory_order_relaxed) != nullptr;
 }
 
-bool Condition::attachConditionVariable(ConditionVariableData* const conditionVariableDataPtr) noexcept
+void Condition::attachConditionVariable(WaitSet* const waitSet,
+                                        ConditionVariableData* const conditionVariableDataPtr) noexcept
 {
     if (isConditionVariableAttached())
     {
-        return false;
+        LogWarn()
+            << "Attaching an already attached condition leads to a detach from the current WaitSet. Best practice "
+               "is to detach Condition first before attaching it.";
+        detachConditionVariable();
     }
 
-    // Call user implementation
-    if (setConditionVariable(conditionVariableDataPtr))
-    {
-        // Save the info so we can notify the user on illegal destruction
-        m_conditionVariableAttached.store(true, std::memory_order_relaxed);
-        return true;
-    }
-    return false;
+    setConditionVariable(conditionVariableDataPtr);
+    m_waitSet.store(waitSet, std::memory_order_relaxed);
 }
 
-bool Condition::detachConditionVariable() noexcept
+void Condition::detachConditionVariable() noexcept
 {
     if (!isConditionVariableAttached())
     {
-        return false;
+        return;
     }
 
-    // Call user implementation
-    if (unsetConditionVariable())
-    {
-        m_conditionVariableAttached.store(false, std::memory_order_relaxed);
-        return true;
-    }
-    return false;
+    unsetConditionVariable();
+    m_waitSet.load(std::memory_order_relaxed)->removeCondition(*this);
```

Comment:
So when a user calls `detachCondition()` on a waitset, now the waitset calls `detachConditionVariable()` on the condition that will call `removeCondition()` on the stored waitset. Puhh

Reply:
Correct but this pull request is the first of some more to come! See: https://github.com/ApexAI/iceoryx/tree/iox-%23341-waitset-next-gen where the behavior will become cleaner soon.

I just would like to have a relatively small pull request as a starting point.
The other approach is that I go completely all in and do the whole thing in one pull request but than it will become massive again :/

Reply:
OK, will have a look again. This is too much ping-pong between the objects